cosmwasm/main.go

```diff
@@ -0,0 +1,280 @@
+						Long:              "This command will install dependencies, create new files, and modify existing files to add a CosmWasm module to the chain.",
+						PlaceCommandUnder: "ignite cosmwasm",
+					},
+				},
+			},
+		},
+	}, nil
+}
+
+func (p *p) Execute(cmd plugin.ExecutedCommand) error {
+
+	// According to the number of declared commands, you may need a switch:
+
+	switch cmd.Use {
+	case "add":
+		return handleAddCommand(p)
+	default:
+		return fmt.Errorf("unknown command: %s", cmd.Use)
+	}
+}
+
+func handleAddCommand(p *p) error {
+	fmt.Println("Adding stuff...")
+
+	err := installDependencies()
+	if err != nil {
+		return err
+	}
+
+	err = createNewFiles(p)
+	if err != nil {
+		return err
+	}
+
+	//err = modifyExistingFiles()
+	//if err != nil {return err}
+
+	return nil
+}
+
+func (p) ExecuteHookPre(hook plugin.ExecutedHook) error {
+	return nil
+}
+
+func (p) ExecuteHookPost(hook plugin.ExecutedHook) error {
+	return nil
+}
+
+func (p) ExecuteHookCleanUp(hook plugin.ExecutedHook) error {
+	return nil
+}
+
+func installDependencies() error {
+	cmd := exec.Command("go", "get", "github.com/CosmWasm/wasmd/x/wasm@v0.41.0")
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+	err := cmd.Run()
+	if err != nil {
+		return fmt.Errorf("failed to install dependency: %w", err)
+	}
+	return nil
+}
+
+func replaceWordsInFile(filePath, targetWord, replacement string) error {
+	// Read content from the file
+	content, err := ioutil.ReadFile(filePath)
+	if err != nil {
+		return err
+	}
+
+	// Perform the replacement
+	modifiedContent := strings.ReplaceAll(string(content), targetWord, replacement)
+
+	// Write the modified content back to the file
+	err = ioutil.WriteFile(filePath, []byte(modifiedContent), 0644)
+	if err != nil {
+		return err
+	}
+
+	//fmt.Printf("Replaced planet with %s\n", replacement)
+
+	return nil
+}
+
+func createFile(inputFilename, outputDir, outputFilename string, p *p) error {
+
+	// Load the embedded template files
+	templatesFS, err := fs.Sub(templates, "wasm-wiring")
+	if err != nil {
+		fmt.Println("Error:", err)
+		return err
+	}
+
+	// Open the embedded input file
+	inputFile, err := templatesFS.Open(inputFilename)
+
+	if err != nil {
+		return err
+	}
+	defer inputFile.Close() // This line schedules the file to be closed when the function returns
+
+	// Read content from the embedded file
+	sourceContent, err := io.ReadAll(inputFile)
+	if err != nil {
+		return err
+	}
+
+	// Construct the output file path
+	outputPath := filepath.Join(outputDir, outputFilename)
+
+	// Write the embedded content to the output file
+	err = ioutil.WriteFile(outputPath, sourceContent, 0644)
+	if err != nil {
+		return err
+	}
+
+	// Replace chainName in the output file
+	err = replaceWordsInFile(outputPath, "planet", p.ChainName())
+	if err != nil {
+		return err
+	}
+
+	fmt.Printf("Created %s in %s\n", outputFilename, outputDir)
+	return nil
+}
+
+func createNewFiles(p *p) error {
```

Comment:
I think here you can just send the `chainName` so you don't pass a complex type like the plugin. That way you keep the function API simpler and easy to unit test.

```suggestion
func createNewFiles(chainName string) error {
```

It will also allow you to remove the `ChainNamed()` method because instead you can just do something like:

```go
outDir:   filepath.Join("cmd", chainName + "d", "cmd"),
```

Also with this change the `createFile()` function will be simpler with only `string` arguments:

```go
err := createFile(f.template, f.outDir, f.outFile, chainName)
```

Reply:
resolved: https://github.com/yerasyla/plugins/commit/27f2900b6c853a13a1269fa3d74fe471399bc399